### DISCUSSION TITLE
Log window implement

### DISCUSSION THREAD

**Discussion by zian31** [emericg/toolBLEx#10] [Ideas]
Hi !
First : What a good job this software, I was looking for a NRFConnect "like" (only on Android) for a while now.
I wonder if it's possible to implement a simple "Log window" with text lines which can notify each news in real time (discover, advertising, etc...).
Thx

**Comment by zian31** [comment]
Or I'm thinking in particular to a timestamp of the Notify (or Indicate) hexa datas with a real time log windows (with export possible), when one (or many) Notify button is(are) ON.

**Reply by emericg** [reply]
I'm glad you enjoy the software :)

Now regarding your question, yes a per device log panel would be useful I think. I'll try to get that in :+1: 
Logging most events (like you said, discover, advertising, services scanned, and the manually triggered events like read and notify) is possible, however the software is not particularly low level when it comes to Bluetooth, so that's probably about it.

**Reply by emericg** [reply]
Would something like that be what you had in mind?
![Capture d’écran du 2023-11-03 19-21-51](https://github.com/emericg/toolBLEx/assets/691258/1417a16b-7a40-4dc6-b16f-624f8a601cfe)

**Reply by zian31** [reply]
Nice new feature with new tab.

If was thinking to the log text tab in NRFConnect per device (with options more or less infos : normal to debug), which is close to your export text :

> Device Name: ATC_90EA05
> Device MAC: A4:C1:38:90:EA:05
> Device MAC manufacturer: Telink Semiconductor (Taipei) Co. Ltd.
> 
> First seen: Wed Nov 1 13:54:55 2023
> Last seen: Fri Nov 3 15:31:05 2023
> 
> Advertising interval: 15966 ms
> Advertisement packet(s):
> > SVD > 15:29:25.262 > 0x181A > ( 6 bytes) 0xa4c13890ea05
> > SVD > 15:27:25.094 > 0xFE95 > (15 bytes) 0x50305b052805ea9038c1a4 0a100146
> ...
> > SVD > 15:21:13.379 > 0x181A > ( 6 bytes) 0xa4c13890ea05
> > SVD > 15:21:13.254 > 0xFE95 > (15 bytes) 0x50305b050e05ea9038c1a4 0a100132
> Service Name: "Generic Access"
> Service UUID: {00001800-0000-1000-8000-00805F9B34FB}
> Characteristic Name: GAP Device Name - UUID: {00002A00-0000-1000-8000-00805F9B34FB} - Properties: Read / Notify - Value: <none>
> Characteristic Name: GAP Appearance - UUID: {00002A01-0000-1000-8000-00805F9B34FB} - Properties: Read - Value: <none>
> Characteristic Name: GAP Peripheral Preferred Connection Parameters - UUID: {00002A04-0000-1000-8000-00805F9B34FB} - Properties: Read - Value: <none>

So a new tab (in the main window of the software) for device with the same infos as your export text in real time (instead of only with export file) would be great (options to see less infos than Debug Mode as in NRFConnect are nice too).
 Thanks :)

**Reply by emericg** [reply]
I've pushed a first version of the log panel. It has a "text" view and an "advanced" view, and it can be saved.

![Capture d’écran du 2023-11-16 18-25-27](https://github.com/emericg/toolBLEx/assets/691258/5f9456ce-5d5f-4fe8-8faa-32610006d83b)
![Capture d’écran du 2023-11-16 18-26-09](https://github.com/emericg/toolBLEx/assets/691258/ffe597c5-7df9-4f67-9b85-ddc97126056f)

**Reply by zian31** [reply]
It seems to be great, it will test it as soon as possible.

**Reply by zian31** [reply]
I don't find this new version, can you publish it here ? Thx
EDIT : the last version in "Actions" seems OK with these new logs
